Refactoring.

swnsma opened the pull request:
Apply code review changes.

See https://github.com/magento/inventory/pull/3088

novikor committed:
Composite products support for IsProductSaleable.

Commit:
Refactoring.
Apply code review changes.

sidolov commented:
@magento run all tests

swnsma commented:
@magento run all tests

Commit:
Fix static tests.

swnsma commented:
@magento run all tests

Commit:
Move product load after type discovery.

swnsma commented:
@magento run all tests

Commit:
Fix typo.

swnsma commented:
@magento run all tests

novikor reviewed (commented):
Looks fine for now. Let's wait for the test builds.

novikor commented:
Well, looks promising (not really)

Commit:
Refactoring.
Move bundle status check to plugin.
Move configurable plugin to around.
Both facts comes from fact for bundle and configurable produts, salability status is depends only on salability status of their children (and does not depend on the condition chain).

swnsma commented:
@magento run all tests

novikor commented on InventoryBundleProduct/Plugin/InventorySales/IsBundleProductSalablePlugin.php:
```diff
@@ -0,0 +1,110 @@
+            }
+
+            // @TODO VERY temporary solution untill https://github.com/magento/inventory/pull/3088 will be resolved
+            // Product salability MUST NOT BE CALLED during product load.
```
I think you refer to a wrong PR. #3088 is actually the current PRs base branch.

novikor commented on InventoryBundleProduct/Plugin/InventorySales/IsBundleProductSalablePlugin.php:
```diff
@@ -0,0 +1,110 @@
+                return $proceed($sku, $stockId);
+            }
+
+            // @TODO VERY temporary solution untill https://github.com/magento/inventory/pull/3088 will be resolved
```
```suggestion
            // @TODO VERY temporary solution until https://github.com/magento/inventory/pull/3088 is resolved
```

novikor commented on InventoryCatalog/Plugin/CatalogInventory/Helper/Stock/AdaptAssignStatusToProductPlugin.php:
```diff
@@ -78,6 +79,17 @@ public function beforeAssignStatusToProduct(
         }
 
         try {
+            // @TODO VERY temporary solution untill https://github.com/magento/inventory/pull/3088 will be resolved
```
The same as above

novikor commented on InventoryBundleProduct/Plugin/InventorySales/IsBundleProductSalablePlugin.php:
```diff
@@ -0,0 +1,110 @@
+/**
+ * Check if bundle product is salable with bundle options.
+ */
+class IsBundleProductSalablePlugin
```
The namespace already includes `Plugin`.
Yes, we already have a bunch of plugins in the Inventory codebase with a similar naming and it is not described in the technical guidelines, but please take attention anyway.

novikor commented on InventoryConfigurableProduct/Plugin/InventorySales/IsConfigurableProductSalablePlugin.php:
```diff
@@ -0,0 +1,124 @@
+ * Check if configurable product is salable.
+ * Moved from chain due to framework limitations (usage of 'AreProductsSalableInterface' provides cyclical dependency).
+ */
+class IsConfigurableProductSalablePlugin
```
The namespace already contains `Plugin`.

novikor reviewed (changes requested):
Looks fine, - but it seems we do not have any active PR and/or issue for disabling `\Magento\CatalogInventory\Observer\AddInventoryDataObserver` (calculating salability status on product load) yet.

Commit:
Fix static tests, fix integration tests, code review fixes, revert unneded changes.

swnsma commented:
@magento run all tests

swnsma commented:
@magento run all tests

novikor committed:
Composite products support for IsProductSaleable.
Fixed case when composite product has out of stock status.

novikor commented:
@magento run Integration Tests

novikor committed:
Merge remote-tracking branch 'origin/composite-products-support-for-isProductSalable' into composite-products-support-for-isProductSalable-v2

novikor commented:
@magento run Integration Tests

swnsma commented:
@magento run all tests

swnsma commented:
@magento run Functional Tests CE

Commit:
magento/inventory#3140:
Fix static tests.

swnsma commented:
@magento run all tests

swnsma commented:
Failed Integration Tests becomes fixed with removal of _Magento\CatalogInventory\Observer\AddInventoryDataObserver_.
As this observer is going to be removal future, my suggestion is to skip this point for now.
